runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1099 @@
+   *
+   * \tparam T Type of the event
+   * \param callback Caller function which accepts the unpacked event
+   * \note: In case type T and actual type do not match an exception is thrown
+   */
+  template <typename T>
+  void subscribe_impl(
+      const std::function<void(databroker::Event<databroker::signal_type_cref_t<T>>&)>& callback) {
+    typed_value_changed_event_t<T>* value_changed_event =
+        std::any_cast<typed_value_changed_event_t<T>>(&value_changed_event_);
+    if (callback) {
+      assert(value_changed_event);
+      callback(*value_changed_event);
+    }
+  }
+
+ public:
+  /**
+   * Subscribe to value-changed events.
+   *
+   * \tparam T Type of the signal
+   * \param callback event-function which will be called when the value changed
+   */
+  template <typename T>
+  void subscribe(databroker::on_value_changed_callback_t<T> callback) {
+    assert_static_type<T>();
+    assert_dynamic_type<T>();
+
+    subscribe_impl<T>(
+        [callback = std::move(callback)](typed_value_changed_event_t<T>& value_changed_event) {
+          value_changed_event.add(std::move(callback));
+        });
+  }
+
+  /**
+   * Return the count of subscribers to the value_changed event.
+   *
+   * \return size_t Count of subscribers to the value_changed event
+   */
+  std::size_t subscriber_count() const { return subscriber_count_(); }
+
+  /**
+   * Indicate whether the value_changed event has subscribers.
+   *
+   * \return bool True if the value_changed event has subscribers, false otherwise
+   */
+  bool has_subscriber() const { return subscriber_count() > 0; }
+
+  /**
+   * Return the list of names assigned to the signal.
+   *
+   * \return List of names assigned to the signal
+   */
+  const std::vector<std::string>& names() const { return names_; }
+
+  /**
+   * Return the first assigned name of the signal.
+   *
+   * \return First name of the signal
+   */
+  const std::string& name() const {
+    if (!names_.size()) {
+      throw std::logic_error("signal does not have a name");
+    }
+    return names_.front();
+  }
+
+  /**
+   * Return the first assigned name of the signal.
+   *
+   * \return First name of the signal
+   */
+  std::string name_or(std::string def) const {
+    if (!names_.size()) {
+      return def;
+    }
+    return names_.front();
+  }
+
+  /**
+   * Add a name of the signal.
+   *
+   * \param name Name of the signal
+   */
+  void add_name(std::string_view name) { names_.emplace_back(name); }
+
+ private:
+  /**
+   * Factory for Signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \return owning unique pointer to Signal
+   * \note: Design-Note #1 reasons the existance of this factory
+   */
+  template <typename T>
+  static signal_ptr_t create() {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    auto signal = std::make_shared<Signal>(access_token(0));
```

Comment:
You know, I think we can get rid of the `access_token` if we just do:
```cpp
auto signal = std::unique_ptr<Signal>(new Signal());
```

The way it's currently implemented, the only thing `access_token` gives us is the ability to use these helper functions, and that only to the class or friends of the class. But the cost to complexity and readability just to avoid using `new` is too high for me. I'd just use new.

```cpp
class Foo {
    struct access_control {
        explicit access_control(int) {}
    };

  public:
    Foo() = delete;
    explicit Foo(access_control) {}

    static std::unique_ptr<Foo> create() {
        return std::make_unique<Foo>(access_control(0));
    }
};

class Bar {
    Bar() = default;

  public:
    static std::unique_ptr<Bar> create() {
        return std::unique_ptr<Bar>(new Bar());
    }
};

int main() {
    auto f1 = Foo::create(); // OK: calling factory method

#if 0
    auto f2 = std::make_unique<Foo>(); // Error
    Foo f3; // Error
#endif

    auto b1 = Bar::create(); // OK: calling factory method

#if 0
    auto b2 = std::make_unique<Bar>(); // Error
    Bar b3; // Error
#endif
}
```

I find Bar a lot easier to read and use.

Reply:
On that note one can think about removing the initialize method, creating event, value changed function, and type in the `create` method to pass them into the ctor. Then the object is directly usable after it's been created.